### PR TITLE
WIP explore using native BigInt

### PR DESCRIPTION
**Pull request by bgrosse-midokura**
~fails with: RangeError: Maximum BigInt size exceeded~

~Changes TypeScript `target` to `esnext`~

Requires a polyfill for BigInt when not available, ~this must be added to readme~ added to README.md

### CODE REVIEW

**Comment by bgrosse-midokura** [comment]
Perhaps this could be fixed by a better `modPow` than `x ** y % z`

File: src/routines.ts [hunk 1]
```diff
@@ -139,16 +131,16 @@ export class SRPRoutines {
...
+    k: bigint,
+    x: bigint,
+    u: bigint,
+    a: bigint,
+    B: bigint,
+  ): bigint {
+    const exp = u * x + a;
+    const tmp = (this.parameters.g ** x % this.parameters.N) * k;
+
+    return (B - tmp) ** exp % this.parameters.N;
```
Comment (bufistov) [inline src/routines.ts:143]: Yess, this should be done in logarithmic way! :)

**Comment by bgrosse-midokura** [comment]
@bufistov fixed with https://www.npmjs.com/package/bigint-mod-arith

There is one test failing: "bigIntegerToWordArray 5 bytes number, big byte", any idea?

**Comment by bufistov** [comment]
> 🍿

My comment was for this [line](https://github.com/midonet/tssrp6a/pull/51/commits/78b8a2eec19002f8096ed9d5ea82c89009e4aab2#diff-51e4f558fae534656963876761c95b83b6ef5da5103c4adef6768219ed76c2deR582)

The test looks stupid, it is safe to remove it I think.

**Comment by bgrosse-midokura** [comment]
Well jsbn probably has the same issue.

**Comment by bufistov** [comment]
> Well jsbn probably has the same issue.

Please check. It might not.

**Comment by bgrosse-midokura** [comment]
@bufistov this is the original source of jsbn: http://www-cs-students.stanford.edu/~tjw/jsbn/

no mention of timing, seems like he was focused on performance, before native BigInt was a thing

**Comment by bufistov** [comment]
> @bufistov this is the original source of jsbn: http://www-cs-students.stanford.edu/~tjw/jsbn/
> 
> no mention of timing, seems like he was focused on performance, before native BigInt was a thing

Yes, no mention of "constant time" and comparison of two numbers is linear loop with "early exit" similar to string comparison.

I am curious now if our current code is vulnerable to timing attacks. It actually might be not. We need to ensure that all functions runtime does not vary depending on the input values. This is easier to to when you have source code of all bigint operations, but I guess the source code of javascript bigint should be also possible to get.

**Comment by bgrosse-midokura** [comment]
@bufistov so, we have the source of jsbn, and native BigInt isn't javascript but the docs say already that it isn't constant. Do you want to analyze jsbn?

**Comment by jbis9051** [comment]
Here's an apparently constant time library https://github.com/soatok/constant-time-js, although I have some doubts about whether it's truly constant time. V8 (and other engines) often optimize JS code. Optimizing code often introduces timing vulnerabilities. 

The best option would probably be assembly code. That would mean writing most of this library (or at least the math stuff) directly in WASM. 

See this article for more information https://www.chosenplaintext.ca/articles/beginners-guide-constant-time-cryptography.html

**Comment by jbis9051** [comment]
As building and maintaining a library in WASM is extremely difficult to do (let alone do *right*), we may just have to accept the existence of timing attacks and choose the best library or native BigInt.

**Comment by bgrosse-midokura** [comment]
Thanks for sharing that link @jbis9051 
If I understand it correctly, it mentions DIV being always vulnerable. SRP uses modulo extensively, which would be using DIV.
I don't exactly follow how it could be used as a timing attack against SRP, like the equality example.

Perhaps we could focus on hardening the equality checks? That seems to be independent of BigInt implementation anyway. I'm referring to step2 of server, and server-verification in step3 of client.

Of course, I might be missing a lot of stuff here :sweat_smile:

**Comment by bufistov** [comment]
Ok. My feeling is  that timing attacks are not feasible for current SRP6a implementation. Each run we generate crypto-secure random numbers (A client, B Server) that are used in the calculations and completely change all the numbers.

For timing attack to be feasible the secret data must be stable, right?

Lets merge this change, it is very nice.

**Comment by bgrosse-midokura** [comment]
I agree that we should merge it. Timing attacks, if any, would be the same for jsbn and native BigInt.

**Pull request merged**